src/topology/homotopy/basic.lean

```diff
@@ -0,0 +1,303 @@
+namespace homotopy_with
+
+section
+
+variables {f₀ f₁ : C(X, Y)} {P : C(X, Y) → Prop}
+
+instance : has_coe_to_fun (homotopy_with f₀ f₁ P) := ⟨_, λ F, F.to_fun⟩
+
+lemma coe_fn_injective : @function.injective (homotopy_with f₀ f₁ P) (I × X → Y) coe_fn :=
+begin
+  rintros ⟨⟨F, _⟩, _⟩ ⟨⟨G, _⟩, _⟩ h,
+  congr' 2,
+end
+
+@[ext]
+lemma ext {F G : homotopy_with f₀ f₁ P} (h : ∀ x, F x = G x) : F = G :=
+coe_fn_injective $ funext h
+
+@[continuity]
+protected lemma continuous (F : homotopy_with f₀ f₁ P) : continuous F := F.continuous_to_fun
+
+@[simp]
+lemma apply_zero (F : homotopy_with f₀ f₁ P) (x : X) : F (0, x) = f₀ x := F.to_fun_zero x
+
+@[simp]
+lemma apply_one (F : homotopy_with f₀ f₁ P) (x : X) : F (1, x) = f₁ x := F.to_fun_one x
+
+@[simp]
+lemma coe_to_continuous_map (F : homotopy_with f₀ f₁ P) : ⇑F.to_continuous_map = F := rfl
+
+/--
+Currying a homotopy to a continuous function fron `I` to `C(X, Y)`.
+-/
+def curry (F : homotopy_with f₀ f₁ P) : C(I, C(X, Y)) := F.to_continuous_map.curry
+
+lemma prop (F : homotopy_with f₀ f₁ P) (t : I) : P (F.curry t) := F.prop' t
+
+@[simp]
+lemma curry_apply (F : homotopy_with f₀ f₁ P) (t : I) (x : X) : F.curry t x = F (t, x) := rfl
+
+/--
+Continuously extending a curried homotopy to a function from `ℝ` to `C(X, Y)`.
+-/
+def extend (F : homotopy_with f₀ f₁ P) : C(ℝ, C(X, Y)) := F.curry.Icc_extend zero_le_one
+
+lemma extend_apply_of_le_zero (F : homotopy_with f₀ f₁ P) {t : ℝ} (ht : t ≤ 0) (x : X) :
+  F.extend t x = f₀ x :=
+begin
+  rw [←F.apply_zero],
+  exact continuous_map.congr_fun (set.Icc_extend_of_le_left (@zero_le_one ℝ _) F.curry ht) x,
+end
+
+lemma extend_apply_of_one_le (F : homotopy_with f₀ f₁ P) {t : ℝ} (ht : 1 ≤ t) (x : X) :
+  F.extend t x = f₁ x :=
+begin
+  rw [←F.apply_one],
+  exact continuous_map.congr_fun (set.Icc_extend_of_right_le (@zero_le_one ℝ _) F.curry ht) x,
+end
+
+@[simp]
+lemma extend_apply_coe (F : homotopy_with f₀ f₁ P) (t : I) (x : X) : F.extend t x = F (t, x) :=
+continuous_map.congr_fun (set.Icc_extend_coe (@zero_le_one ℝ _) F.curry t) x
+
+@[simp]
+lemma extend_apply_of_mem_I (F : homotopy_with f₀ f₁ P) {t : ℝ} (ht : t ∈ I) (x : X) :
+  F.extend t x = F (⟨t, ht⟩, x) :=
+continuous_map.congr_fun (set.Icc_extend_of_mem (@zero_le_one ℝ _) F.curry ht) x
+
+lemma extend_prop (F : homotopy_with f₀ f₁ P) (t : ℝ) : P (F.extend t) :=
+begin
+  by_cases ht₀ : 0 ≤ t,
+  { by_cases ht₁ : t ≤ 1,
+    { convert F.prop ⟨t, ht₀, ht₁⟩,
+      ext,
+      rw [F.extend_apply_of_mem_I ⟨ht₀, ht₁⟩, F.curry_apply] },
+    { convert F.prop 1,
+      ext,
+      rw [F.extend_apply_of_one_le (le_of_not_le ht₁), F.curry_apply, F.apply_one] } },
+  { convert F.prop 0,
+    ext,
+    rw [F.extend_apply_of_le_zero (le_of_not_le ht₀), F.curry_apply, F.apply_zero] }
+end
+
+end
+
+variable {P : C(X, Y) → Prop}
+
+/--
+Given a continuous function `f` and a proof `hP : P f`, we can define a `homotopy_with f f` by
+`F (t, x) = f x`
+-/
+def refl (f : C(X, Y)) (hP : P f) : homotopy_with f f P :=
```

Comment:
This should probably be tagged with `@[simps]` to autogenerate a `refl_apply` lemma.

Reply:
`@[simps]` should be able to do this, but it needs some setup when extending a structure without using `old_structure_cmd`, see for example here: https://github.com/leanprover-community/mathlib/blob/540fb9434673e774031b108643d8f47dcb6438ab/src/topology/local_homeomorph.lean#L83-L91

Reply:
🏎️, you beat me to it, oops.

Reply:
Thanks!